apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/DroppedSpanStats.java

```diff
@@ -78,23 +86,35 @@ public Outcome getOutcome() {
 
         @Override
         public void resetState() {
-            destinationServiceResource = null;
             serviceType = null;
-            serviceName = null;
-            outcome = null;
+            serviceName.setLength(0);
+            destinationResource.setLength(0);
+            outcome = Outcome.UNKNOWN;
         }
 
         @Override
         public boolean equals(Object o) {
             if (this == o) return true;
             if (o == null || getClass() != o.getClass()) return false;
+
             StatsKey statsKey = (StatsKey) o;
-            return serviceType.equals(statsKey.serviceType) && Objects.equals(serviceName, statsKey.serviceName) && outcome == statsKey.outcome;
+
+            Objects.requireNonNull(serviceType);
+
+            if (!serviceType.equals(statsKey.serviceType)) return false;
+            if (!CharSequenceUtils.equals(serviceName, statsKey.serviceName)) return false;
+            if (!CharSequenceUtils.equals(destinationResource, statsKey.destinationResource)) return false;
+            return outcome == statsKey.outcome;
         }
 
         @Override
         public int hashCode() {
-            return Objects.hash(serviceType, serviceName, outcome);
+            Objects.requireNonNull(serviceType);
+            int result = serviceType.hashCode();
+            result = 31 * result + CharSequenceUtils.hashCode(serviceName);
+            result = 31 * result + CharSequenceUtils.hashCode(destinationResource);
+            result = 31 * result + outcome.hashCode();
+            return result;
         }
```

Comment:
Quite heavy calculations for that, but looks it was similar with `Objects.hash`. Let's cache the hash so that it is only calculated once (and clear on `reset`)